File: cpp/tests/scheme_tests.cpp
#include "packbin/packbin.hpp"

#include <fstream>
#include <iostream>
#include <sstream>
#include <string>
#include <vector>

namespace {

int failures = 0;

void expect(bool ok, char const* msg) {
  if (!ok) {
    std::cerr << "FAIL: " << msg << "\n";
    ++failures;
  }
}

std::vector<std::uint8_t> parse_hex(std::string hex) {
  while (!hex.empty() && (hex.back() == '\n' || hex.back() == '\r' || hex.back() == ' '))
    hex.pop_back();
  std::vector<std::uint8_t> out(hex.size() / 2);
  for (std::size_t i = 0; i < out.size(); ++i)
    out[i] = static_cast<std::uint8_t>(std::stoul(hex.substr(i * 2, 2), nullptr, 16));
  return out;
}

std::string find_golden() {
  char const* candidates[] = {
      "../fixtures/golden.hex",
      "fixtures/golden.hex",
      "../../fixtures/golden.hex",
  };
  for (auto const* path : candidates) {
    std::ifstream in(path);
    if (in) {
      std::stringstream buf;
      buf << in.rdbuf();
      return buf.str();
    }
  }
  return {};
}

std::string read_file(char const* path) {
  std::ifstream in(path);
  if (!in)
    return {};
  std::stringstream buf;
  buf << in.rdbuf();
  return buf.str();
}

std::string api_sources() {
  char const* roots[] = {
      "include/packbin/",
      "cpp/include/packbin/",
      "../include/packbin/",
  };
  for (auto const* root : roots) {
    auto main = read_file((std::string(root) + "packbin.hpp").c_str());
    auto scheme = read_file((std::string(root) + "scheme.hpp").c_str());
    if (!main.empty() && !scheme.empty())
      return main + "\n" + scheme;
  }
  return {};
}

struct MarkerRow {
  std::uint8_t sid = 0;
};

struct PositionRow {
  std::uint16_t sid = 0;
  std::int32_t lat = 0;
  std::int32_t lon = 0;
  std::uint8_t profile = 0;
};

struct UserModifiedEvent {
  std::int32_t user_id = 0;
  std::string user_name_change;
  std::string user_email_change;
  std::uint8_t user_status_change = 0;
};

struct UserPositionEvent {
  std::int32_t user_id = 0;
  std::int32_t latitude = 0;
  std::int32_t longitude = 0;
};

auto position_scheme() {
  return packbin::Scheme<PositionRow>(
      0x40, packbin::u16(0, &PositionRow::sid), packbin::i32(1, &PositionRow::lat),
      packbin::i32(2, &PositionRow::lon), packbin::u8(3, &PositionRow::profile),
      packbin::flags({packbin::u16(4), packbin::u8(5), packbin::i16(6)}));
}

auto modified_scheme() {
  return packbin::Scheme<UserModifiedEvent>(
      1, packbin::i32(0, &UserModifiedEvent::user_id),
      packbin::utf8(1, &UserModifiedEvent::user_name_change),
      packbin::utf8(2, &UserModifiedEvent::user_email_change),
      packbin::u8(3, &UserModifiedEvent::user_status_change));
}

auto position_event_scheme() {
  return packbin::Scheme<UserPositionEvent>(
      2, packbin::i32(0, &UserPositionEvent::user_id),
      packbin::i32(1, &UserPositionEvent::latitude),
      packbin::i32(2, &UserPositionEvent::longitude));
}

void ac1_scheme_replaces_packet() {
  auto source = api_sources();
  expect(!source.empty(), "AC-1 header found");
  expect(source.find("struct Packet") == std::string::npos, "AC-1 no struct Packet");
  expect(source.find("struct BinaryPacker") != std::string::npos, "AC-1 BinaryPacker");
  expect(source.find("TypeNum") == std::string::npos, "AC-1 no TypeNum");
  expect(source.find("type_num(") == std::string::npos, "AC-1 no type_num");
  packbin::Scheme<MarkerRow> s(32, packbin::u8(0, &MarkerRow::sid));
  expect(s.type_number == 32, "AC-1 type_number");
  MarkerRow row;
  row.sid = 23;
  auto bytes = packbin::BinaryPacker::pack(s, row);
  expect(packbin::to_hex(bytes) == "2017", "AC-1 pack hex");
}

void ac2_position_golden_no_type_member() {
  PositionRow row;
  row.sid = 1;
  row.lat = 500000000;
  row.lon = 300000000;
  row.profile = 1;
  auto bytes = packbin::BinaryPacker::pack(position_scheme(), row);
  expect(packbin::to_hex(bytes) == "4001000065cd1d00a3e1110100", "AC-2 golden hex");
  auto fixture = find_golden();
  expect(!fixture.empty(), "AC-2 golden.hex found");
  expect(packbin::mismatched_bytes(bytes, parse_hex(fixture)) == 0, "AC-2 fixture");
  auto source = read_file("tests/scheme_tests.cpp");
  if (source.empty())
    source = read_file("cpp/tests/scheme_tests.cpp");
  expect(!source.empty(), "AC-2 source found");
  auto start = source.find("struct PositionRow");
  expect(start != std::string::npos, "AC-2 PositionRow present");
  auto end = source.find("};", start);
  expect(end != std::string::npos, "AC-2 PositionRow body");
  auto body = source.substr(start, end - start);
  expect(body.find("type") == std::string::npos, "AC-2 no type member");
}

void ac3_known_scheme_checks_leading_byte() {
  auto layout = packbin::Scheme<MarkerRow>(1, packbin::u8(0, &MarkerRow::sid));
  auto got = packbin::BinaryPacker::unpack(layout, parse_hex("0217"));
  expect(!got.ok, "AC-3 not ok");
  expect(!got.value.has_value(), "AC-3 no row");
  expect(got.type_mismatch.has_value(), "AC-3 type_mismatch");
  if (got.type_mismatch) {
    expect(got.type_mismatch->expected.has_value() && *got.type_mismatch->expected == 1,
           "AC-3 expected 1");
    expect(got.type_mismatch->actual == 2, "AC-3 actual 2");
  }
}

void ac4_unknown_buffer_matching_handler() {
  UserPositionEvent row;
  row.user_id = 7;
  row.latitude = 8;
  row.longitude = 9;
  auto bytes = packbin::BinaryPacker::pack(position_event_scheme(), row);
  expect(packbin::to_hex(bytes) == "02070000000800000009000000", "AC-4 hex");

  int modified = 0;
  int position = 0;
  UserPositionEvent seen{};
  auto result = packbin::BinaryPacker::unpack(
      bytes, modified_scheme().on([&](UserModifiedEvent const&) { ++modified; }),
      position_event_scheme().on([&](UserPositionEvent const& ev) {
        ++position;
        seen = ev;
      }));
  expect(result.ok, "AC-4 ok");
  expect(modified == 0, "AC-4 modified not called");
  expect(position == 1, "AC-4 position called");
  expect(seen.user_id == 7, "AC-4 user_id");
  expect(seen.latitude == 8, "AC-4 latitude");
  expect(seen.longitude == 9, "AC-4 longitude");
}

void ac5_unknown_type_number() {
  int modified = 0;
  int position = 0;
  auto result = packbin::BinaryPacker::unpack(
      std::vector<std::uint8_t>{9},
      modified_scheme().on([&](UserModifiedEvent const&) { ++modified; }),
      position_event_scheme().on([&](UserPositionEvent const&) { ++position; }));
  expect(!result.ok, "AC-5 not ok");
  expect(result.type_mismatch.has_value(), "AC-5 type_mismatch");
  if (result.type_mismatch) {
    expect(!result.type_mismatch->expected.has_value(), "AC-5 no expected");
    expect(result.type_mismatch->actual == 9, "AC-5 actual 9");
  }
  expect(modified == 0 && position == 0, "AC-5 no handlers");
}

void ac6_type_numbers_unique() {
  bool threw = false;
  try {
    packbin::BinaryPacker::unpack(std::vector<std::uint8_t>{1},
                    modified_scheme().on([](UserModifiedEvent const&) {}),
                    packbin::Scheme<UserModifiedEvent>(1, packbin::i32(0, &UserModifiedEvent::user_id))
                        .on([](UserModifiedEvent const&) {}));
  } catch (std::runtime_error const&) {
    threw = true;
  }
  expect(threw, "AC-6 duplicate throws");
}

void type_number_range() {
  bool high = false;
  bool low = false;
  try {
    packbin::scheme(256, {packbin::u8(0)});
  } catch (std::runtime_error const&) {
    high = true;
  }
  try {
    packbin::scheme(-1, {packbin::u8(0)});
  } catch (std::runtime_error const&) {
    low = true;
  }
  bool typed_high = false;
  try {
    packbin::Scheme<MarkerRow>(256, packbin::u8(0, &MarkerRow::sid));
  } catch (std::runtime_error const&) {
    typed_high = true;
  }
  expect(high && low && typed_high, "type number 0..255");
}

}  // namespace

int run_scheme_tests() {
  ac1_scheme_replaces_packet();
  ac2_position_golden_no_type_member();
  ac3_known_scheme_checks_leading_byte();
  ac4_unknown_buffer_matching_handler();
  ac5_unknown_type_number();
  ac6_type_numbers_unique();
  type_number_range();
  return failures;
}

File: csharp/tests/SchemeTests.cs
using System.Diagnostics;
using Packbin;

namespace Packbin.Tests;

public class SchemeTests
{
    public sealed class MarkerRow
    {
        public byte Sid { get; set; }
    }

    public sealed class UserModifiedEvent
    {
        public int UserId { get; set; }
        public string UserNameChange { get; set; } = "";
        public string UserEmailChange { get; set; } = "";
        public byte UserStatusChange { get; set; }
    }

    public sealed class UserPositionEvent
    {
        public int UserId { get; set; }
        public int Latitude { get; set; }
        public int Longitude { get; set; }
    }

    public sealed class PositionRow
    {
        public ushort Sid { get; set; }
        public int Lat { get; set; }
        public int Lon { get; set; }
        public byte Profile { get; set; }
        public ushort? Heading { get; set; }
        public byte? Speed { get; set; }
        public short? Altitude { get; set; }
    }

    private static readonly Scheme<MarkerRow> MarkerScheme = new(32, f => [f.U8(0, x => x.Sid)]);

    private static readonly Scheme<UserModifiedEvent> ModifiedScheme = new(1, f => [
        f.I32(0, x => x.UserId),
        f.Utf8(1, x => x.UserNameChange),
        f.Utf8(2, x => x.UserEmailChange),
        f.U8(3, x => x.UserStatusChange)]);

    private static readonly Scheme<UserPositionEvent> PositionEventScheme = new(2, f => [
        f.I32(0, x => x.UserId),
        f.I32(1, x => x.Latitude),
        f.I32(2, x => x.Longitude)]);

    private static readonly Scheme<PositionRow> PositionScheme = new(0x40, f => [
        f.U16(0, x => x.Sid),
        f.I32(1, x => x.Lat),
        f.I32(2, x => x.Lon),
        f.U8(3, x => x.Profile),
        f.Flags(
            f.U16(4, x => x.Heading),
            f.U8(5, x => x.Speed),
            f.I16(6, x => x.Altitude))]);

    private static readonly Dictionary<string, object?> PositionValues = new()
    {
        ["Sid"] = (ushort)1,
        ["Lat"] = 500_000_000,
        ["Lon"] = 300_000_000,
        ["Profile"] = (byte)1,
    };

    private const string GoldenHex = "4001000065cd1d00a3e1110100";
    private const string Ac4Hex = "02070000000800000009000000";

    [Fact]
    public void Ac1_SchemeReplacesPacket()
    {
        Assert.Null(typeof(Packbin.Scheme<MarkerRow>).Assembly.GetType("Packbin.Packet"));
        Assert.NotNull(typeof(Packbin.Scheme<MarkerRow>).Assembly.GetType("Packbin.BinaryPacker"));
        Assert.Null(typeof(Packbin.Scheme<MarkerRow>).Assembly.GetType("Packbin.TypeNum"));
        var scheme = new Scheme<MarkerRow>(32, Field.U8<MarkerRow>(0, x => x.Sid));
        Assert.Equal(32, scheme.TypeNumber);
        Assert.Single(scheme.Fields);
        Assert.ThrowsAny<ArgumentException>(() => new Scheme<MarkerRow>(-1, Field.U8<MarkerRow>(0, x => x.Sid)));
        Assert.ThrowsAny<ArgumentException>(() => new Scheme<MarkerRow>(256, Field.U8<MarkerRow>(0, x => x.Sid)));
        var bytes = BinaryPacker.Pack(scheme, new MarkerRow { Sid = 23 });
        Assert.Equal("2017", Convert.ToHexString(bytes).ToLowerInvariant());
    }

    [Fact]
    public void Ac2_RowHasNoTypeMember()
    {
        var bytes = BinaryPacker.Pack(PositionScheme, PositionValues);
        Assert.Equal(GoldenHex, Convert.ToHexString(bytes).ToLowerInvariant());
        var fixture = ParseHex(File.ReadAllText(FindGoldenFixture()).Trim());
        Assert.Equal(0, MismatchedBytes(bytes, fixture));
        Assert.Null(typeof(PositionRow).GetProperty("type"));
        Assert.Null(typeof(PositionRow).GetField("type"));
        PositionRow? row = null;
        var err = BinaryPacker.Unpack(bytes, PositionScheme.On(v => row = v));
        Assert.Null(err);
        Assert.NotNull(row);
        Assert.Equal((ushort)1, row.Sid);
        Assert.Equal(500_000_000, row.Lat);
        Assert.Equal(300_000_000, row.Lon);
        Assert.Equal((byte)1, row.Profile);
    }

    [Fact]
    public void Ac3_KnownSchemeChecksLeadingByte()
    {
        var scheme = new Scheme<MarkerRow>(1, Field.U8<MarkerRow>(0, x => x.Sid));
        MarkerRow? row = null;
        var err = BinaryPacker.Unpack(ParseHex("0217"), scheme.On(v => row = v));
        Assert.Null(row);
        var mismatch = Assert.IsType<TypeMismatch>(err);
        Assert.Equal(2, mismatch.Actual);
    }

    [Fact]
    public void Ac4_UnknownBufferCallsMatchingHandler()
    {
        UserModifiedEvent? modified = null;
        UserPositionEvent? position = null;
        var err = BinaryPacker.Unpack(
            ParseHex(Ac4Hex),
            ModifiedScheme.On(ev => modified = ev),
            PositionEventScheme.On(ev => position = ev));
        Assert.Null(err);
        Assert.Null(modified);
        Assert.NotNull(position);
        Assert.Equal(7, position.UserId);
        Assert.Equal(8, position.Latitude);
        Assert.Equal(9, position.Longitude);
        Assert.Null(typeof(UserPositionEvent).GetProperty("type"));
        Assert.Null(typeof(UserPositionEvent).GetField("type"));
        var packed = BinaryPacker.Pack(PositionEventScheme, new UserPositionEvent
        {
            UserId = 7,
            Latitude = 8,
            Longitude = 9,
        });
        Assert.Equal(Ac4Hex, Convert.ToHexString(packed).ToLowerInvariant());
    }

    [Fact]
    public void Ac5_UnknownTypeNumber()
    {
        var modifiedRan = false;
        var positionRan = false;
        var err = BinaryPacker.Unpack(
            ParseHex("09070000000800000009000000"),
            ModifiedScheme.On(_ => modifiedRan = true),
            PositionEventScheme.On(_ => positionRan = true));
        Assert.False(modifiedRan);
        Assert.False(positionRan);
        var mismatch = Assert.IsType<TypeMismatch>(err);
        Assert.Equal(9, mismatch.Actual);
    }

    [Fact]
    public void Ac6_TypeNumbersInOneCallAreUnique()
    {
        var other = new Scheme<UserModifiedEvent>(1,
            Field.I32<UserModifiedEvent>(0, x => x.UserId),
            Field.Utf8<UserModifiedEvent>(1, x => x.UserNameChange),
            Field.Utf8<UserModifiedEvent>(2, x => x.UserEmailChange),
            Field.U8<UserModifiedEvent>(3, x => x.UserStatusChange));
        Assert.ThrowsAny<ArgumentException>(() =>
            BinaryPacker.Unpack(
                ParseHex(Ac4Hex),
                ModifiedScheme.On(_ => { }),
                other.On(_ => { })));
    }

    [Fact]
    public void EmptyBufferIsShortPacket()
    {
        MarkerRow? row = null;
        var err = BinaryPacker.Unpack(ReadOnlySpan<byte>.Empty, MarkerScheme.On(v => row = v));
        Assert.Null(row);
        var missing = Assert.IsType<ShortPacket>(err);
        Assert.Equal("", missing.Field);
        Assert.Equal(1, missing.Needed);
        Assert.Equal(0, missing.Left);
    }

    [Fact]
    public void CompileFailRequiresScheme()
    {
        var fixtureDir = FindCompileFailDir();
        var project = Path.Combine(fixtureDir, "PackWithoutScheme.csproj");
        Assert.True(File.Exists(project), project);
        var start = new ProcessStartInfo
        {
            FileName = "dotnet",
            Arguments = $"build \"{project}\" --nologo -v q",
            RedirectStandardOutput = true,
            RedirectStandardError = true,
            UseShellExecute = false,
        };
        using var process = Process.Start(start) ?? throw new InvalidOperationException("dotnet start failed");
        var stdout = process.StandardOutput.ReadToEnd();
        var stderr = process.StandardError.ReadToEnd();
        process.WaitForExit();
        Assert.True(
            process.ExitCode != 0,
            $"expected compile failure, exit={process.ExitCode}\n{stdout}\n{stderr}");
    }

    private static int MismatchedBytes(byte[] actual, byte[] expected)
    {
        var mismatches = Math.Abs(actual.Length - expected.Length);
        var shared = Math.Min(actual.Length, expected.Length);
        for (var i = 0; i < shared; i++)
        {
            if (actual[i] != expected[i])
                mismatches++;
        }
        return mismatches;
    }

    private static byte[] ParseHex(string hex)
    {
        hex = hex.Trim();
        var bytes = new byte[hex.Length / 2];
        for (var i = 0; i < bytes.Length; i++)
            bytes[i] = Convert.ToByte(hex.Substring(i * 2, 2), 16);
        return bytes;
    }

    private static string FindGoldenFixture()
    {
        var dir = new DirectoryInfo(AppContext.BaseDirectory);
        while (dir is not null)
        {
            var path = Path.Combine(dir.FullName, "fixtures", "golden.hex");
            if (File.Exists(path))
                return path;
            dir = dir.Parent;
        }
        throw new FileNotFoundException("fixtures/golden.hex");
    }

    private static string FindCompileFailDir()
    {
        var dir = new DirectoryInfo(AppContext.BaseDirectory);
        while (dir is not null)
        {
            var path = Path.Combine(dir.FullName, "csharp", "tests", "compile-fail");
            if (Directory.Exists(path))
                return path;
            path = Path.Combine(dir.FullName, "tests", "compile-fail");
            if (Directory.Exists(path))
                return path;
            dir = dir.Parent;
        }
        throw new DirectoryNotFoundException("csharp/tests/compile-fail");
    }
}
